cypress/integration/x-msc2716-historical-import/historical-import.spec.ts

```diff
@@ -0,0 +1,628 @@
+            msgtype: "m.text",
+        });
+
+        // Wait for the message to show up for the logged in user
+        waitForEventIdsInClient([eventIdAfterHistoricalImport]);
+    });
+
+    sendMarkerEventAndEnsureHistoryDetectedStatusBar(asMatrixClient);
+}
+
+describe("MSC2716: Historical Import", () => {
+    let synapse: SynapseInstance;
+    let asMatrixClient: MatrixClient;
+    const virtualUserIDs = ['@maria-01234:localhost'];
+
+    // This corresponds to the application service registration defined in the
+    // "msc2716-historical-import" Synapse configuration
+    const AS_TOKEN = 'as_token123';
+
+    beforeEach(() => {
+        cy.window().then(win => {
+            // Collapse left panel for these tests (get more space in the area we care about)
+            win.localStorage.setItem("mx_lhs_size", "0");
+        });
+        // Start Synapse with msc2716_enabled and an application service configured
+        cy.startSynapse("msc2716-historical-import").then(data => {
+            synapse = data;
+
+            // This is the person we're logged in as
+            cy.initTestUser(synapse, "Grace");
+
+            // After the page is loaded from initializing the logged in user so
+            // the mxSettingsStore is available
+            cy.window().then(win => {
+                // Disable the sound notifications so you're not startled and
+                // confused where the sound is coming from
+                win.mxSettingsStore.setValue("audioNotificationsEnabled", null, SettingLevel.DEVICE, false);
+                // Show hidden events to make debugging easier. And the state
+                // events will show up in the timeline to make it the same
+                // assert as other events.
+                win.mxSettingsStore.setValue("showHiddenEventsInTimeline", null, SettingLevel.DEVICE, true);
+            });
+
+            // Get a Matrix Client for the application service
+            cy.newMatrixClient(synapse, {
+                userId: '@gitter-badger:localhost',
+                accessToken: AS_TOKEN,
+            }).then(matrixClient => {
+                asMatrixClient = matrixClient;
+            });
+
+            ensureVirtualUsersRegistered(synapse, AS_TOKEN, virtualUserIDs);
+        });
+    });
+
+    afterEach(() => {
+        cy.stopSynapse(synapse);
+    });
+
+    it("Shows historical messages after refreshing the timeline", () => {
+        setupRoomWithHistoricalMessagesAndMarker({
+            synapse,
+            asMatrixClient,
+            virtualUserIDs,
+        });
+
+        // Press "Refresh timeline"
+        cy.get(`[data-cy="refresh-timeline-button"]`).click();
+
+        // Ensure historical messages are now shown
+        cy.all([
+            cy.get<string[]>("@liveMessageEventIds"),
+            cy.get<string[]>("@historicalEventIds"),
+        ]).then(([liveMessageEventIds, historicalEventIds]) => {
+            // FIXME: Assert that they appear in the correct order
```

Comment:
We can tackle this in another iteration. It's pretty much good enough to assert that they are on the page. The server is determining the order anyway.